File: BAO/Scripts/BAO/BAO3/BAO3.pl
#!/usr/bin/perl
use utf8;
binmode(STDOUT, ":utf8");

my $file=shift @ARGV; # input
my @PATTERN=@ARGV; # patron morphosyntaxique
open my $input, '<:encoding(utf-8)', $file;
my @LIST=<$input>; # mettre fichier dans liste

close ($input);

while (my $line=shift @LIST) {

=pod
	check that line in $line corresponds to first lexico-syntactic pattern $PATTERN
	then read as many lines as there are in $PATTERN
=cut
	my $term="";

	if ($line=~/<element><data type="type">$PATTERN[0]<\/data><data type="lemma">[^<]+?<\/data><data type="string">([^<]+?)<\/data><\/element>/) {

		$term=$term.$1; # on récupère la forme grâce à une expression régulière qui utilise le patron n°1
		my $index=1;
		my $length=1;
		
		while (($LIST[$index-1]=~/<element><data type="type">($PATTERN[$index])<\/data><data type="lemma">[^<]+?<\/data><data type="string">([^<]+?)<\/data><\/element>/) and ($index <= $#PATTERN)) {

			$term.=" ".$2; # on récupère la forme qui correspond au patron d'après
			$index++;
			$length++;

		}

		if ($length==$#PATTERN+1) {
			$dictPattern{$term}++; # on ajoute le terme qui match à un dictionnaire
			$nbTerm++;
		}
	
	}

}

open my $outfile, ">:encoding(UTF-8)", "perlIsTheBigOne.txt"; # on ouvre un fichier output
print $outfile "$nbTerm éléments trouvés\n"; # on envoie le nombre de termes

foreach my $pattern (sort {$dictPattern{$b} <=> $dictPattern{$a} } keys %dictPattern) {
	print $outfile "$dictPattern{$pattern}\t$pattern\n"; # on envoie également les formes qui correspondent aux patrons
}

close($outfile);

File: BAO/Scripts/BAO/BAO3/BAO3.py
from typing import List
import re
import sys
from pathlib import Path
import time

def extract(file: str, pattern: List[str]): #fonction alternative extract()

	buffer = [("---", "---")] * len(pattern) # buffer – window that slides over entire file by group of the size of the length of the pattern

	with open(file) as corpus:
		for line in corpus:

			buffer.pop(0) # on enlève le premier élément du buffer
			match = re.match('<element><data type="type">([^<]+?)</data><data type="lemma">[^<]+?</data><data type="string">([^<]+?)</data></element>', line)

			if match:
				tag = match.group(1) # on récupère les patrons
				form = match.group(2) # et leurs formes grâce à l'expression régulière match
				buffer.append((tag,form)) # on les ajoute à notre buffer
			else:
				buffer = [("---", "---")] * len(pattern)

			check = True
			term = ""

			for i, tag in enumerate(pattern): # on récupère le patron et son index
				if tag == buffer[i][0]:
					term = term + buffer[i][1] + f"/{tag} " # on ajoute la nouvelle forme qui correspond à notre variable terme
				else:
					check = False # sinon on sort de la boucle, le patron ne correspond pas

			if check:
				print(term) # si le patron correspond, on l'affiche sur le terminal

if __name__ == '__main__':

	start_time = time.time()
	
	file = sys.argv[1] # premier argument est notre fichier XML
	pattern = sys.argv[2:] # second argument et plus est notre patron

	extract(file, pattern)

	print("Running time : %ss" % (time.time() - start_time))
